lizaku/assignment4/assignment4.py

```diff
@@ -0,0 +1,86 @@
+import itertools
+
+class Grid:
+    '''
+    Grid is initially a 2D array, but I turn all values into Node objects
+    and augment them with their neighbors
+    '''
+    def __init__(self, data):
+        # if the user provides the grid that is not valid
+        
+        # turn values into Node objects
+        i = 0
+        for row in range(len(data)):
+            for elem in range(len(data[row])):
+                node = Node(data[row][elem], i, set())
+                data[row][elem] = node
+                i += 1
+        try:        
+            # compute neighbors
+            for row in range(len(data)):
+                for ind in range(len(data[row])):
+                    tile = data[row][ind]
+                    # dinf up neighbor
+                    if row != 0:
+                        tile.neighbors.add(data[row - 1][ind])
+                    # find left neighbor
+                    if ind != 0:
+                        tile.neighbors.add(data[row][ind - 1])
+                    # find right neighbor
+                    if ind != len(data[row]) - 1:
+                        tile.neighbors.add(data[row][ind + 1])
+                    # find down neighbor
+                    if row != len(data) - 1:
+                        tile.neighbors.add(data[row + 1][ind])
+            self.data = data
+        except IndexError:
+            raise IndexError('Different number of tiles in rows')
+    
+    def find_islands(self):
+        islands = []
```

Comment:
Cool explanation of the algorithm! Can you please summarize it at the top so that the person reading can prepare him/herself for what is coming?